### PR TITLE
Added programId field in TokenBalance type

### PR DESCRIPTION
**Pull request by shivaji43**
Tackles the issue (#3091) 
Screenshot 2024-11-19 234833
Screenshot 2024-11-19 233507
Screenshot 2024-11-19 233442

Done proper linting , built and complied properly , also tested to see if programId returns as expected ,
also added the screenshots

Please tell me if I am missing something and Thank you for being patient and helping me through this.

### CODE REVIEW

**Review by steveluscher — CHANGES_REQUESTED** [review]
Adding this property to the TypeScript type is insufficient to vend it to the caller.

Look here:

https://github.com/solana-labs/solana-web3.js/blob/maintenance/v1.x/src/connection.ts#L2329-L2334

To get your PRs in faster, make sure to test your changes, to write down in your PR description exactly what you did, and to paste in the results as proof that the code works.

File: src/connection.ts [hunk 1]
```diff
@@ -2330,6 +2331,7 @@ const TokenBalanceResult = pick({
   accountIndex: number(),
   mint: string(),
   owner: optional(string()),
+  programId:optional(string()),
```
Comment (steveluscher) [inline src/connection.ts:2334]: I can see that you haven't run any tests or build scripts, because this line of code isn't formatted.

**Review by steveluscher — CHANGES_REQUESTED** [review]
giphy

> To get your PRs in faster, make sure to test your changes, to write down in your PR description exactly what you did, and to paste in the results as proof that the code works.

Doing this will make it really easy for me to accept your PR with high confidence.

1. Run code quality scripts

        pnpm test:lint && pnpm test:typecheck && pnpm test:prettier

2. Build the code

        pnpm compile:js && pnpm compile:typedefs

3. Start a REPL

        pnpm node
        > const { Connection } = require('./lib/index.cjs.js');
        > const connection = new Connection('https://api.devnet.solana.com');

4. Do your testing. Use the `connection` instance to make a query, show that the `programId` field materializes on the response, and that there are no fatals. Paste all of your steps and results into the PR description

**Review by steveluscher — CHANGES_REQUESTED** [review]
Almost!

The data structure you edited is called `TokenBalance`

https://github.com/solana-labs/solana-web3.js/blob/maintenance/v1.x/src/connection.ts#L1049-L1054

You can see it in use here:

https://github.com/solana-labs/solana-web3.js/blob/maintenance/v1.x/src/connection.ts#L1085-L1088

And here:

https://github.com/solana-labs/solana-web3.js/blob/maintenance/v1.x/src/connection.ts#L1116-L1119

Those properties are materialized here:

https://github.com/solana-labs/solana-web3.js/blob/maintenance/v1.x/src/connection.ts#L2344-L2370

And here:

https://github.com/solana-labs/solana-web3.js/blob/maintenance/v1.x/src/connection.ts#L2375-L2395

Look at all of the RPC calls that materialize `ConfirmedTransactionMetaResult` and `ParsedConfirmedTransactionMetaResult`.

You'll see that one such call is this one:

https://github.com/solana-labs/solana-web3.js/blob/maintenance/v1.x/src/connection.ts#L2573C7-L2583

So find out where that is used:

https://github.com/solana-labs/solana-web3.js/blob/maintenance/v1.x/src/connection.ts#L5053-L5083

And there we have an actual RPC call you can make; `getTransaction()`. There are many others, but this is one. You could conceive of a transaction that you could get, then check `result.meta.postTokenBalances` to see if you got the new property!

Just make sure you actually use a transaction that involves SPL tokens, or that array will be empty and you won't learn anything.

**Comment by shivaji43** [comment]
Screenshot 2024-11-20 152521

I got it , I put one of my own transaction signature which involved actual token transfer  and i was able to the programId in the metadata array now.

**Comment by steveluscher** [comment]
Beautiful. Thank you!

**Comment by shivaji43** [comment]
Thank you so much @steveluscher

**Pull request merged**